### PR TITLE
Fix clonerefs duplicate ref bug and add unit tests.

### LINKED ISSUE #10093: Jobs cloning the kubernetes/release repo are failing due to `/home/prow/go/src/k8s.io/release/.git/hooks/post-update.sample': File exists # Error: exit status 128
Example: https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-coverage-conformance/404/build-log.txt
```
# FAILED!
# Cloning kubernetes/release at master
$ mkdir -p /home/prow/go/src/k8s.io/release
$ git init
fatal: cannot copy '/usr/share/git-core/templates/hooks/post-update.sample' to '/home/prow/go/src/k8s.io/release/.git/hooks/post-update.sample': File exists
# Error: exit status 128
# FAILED!
```

/cc @krzyzacy

### PR DESCRIPTION
**Pull request by cjwagner**
fixes #10093 
ref #9952 

I just added a couple simple unit unit tests to validate the fix so that we can get this fix out quickly. It would be great to improve test coverage later.
/assign @fejta 
/cc @BenTheElder @yujuhong

### CODE REVIEW

**Comment by cjwagner** [comment]
We'll need to bump Prow to pick up this fix after it merges.

File: prow/clonerefs/run_test.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+		refs        *kube.Refs
+		root        string
+		user, email string
+		cookiePath  string
+		env         []string
+	}
+
+	var recordedClones []cloneRec
+	cloneFuncOld := cloneFunc
+	cloneFunc = func(refs *kube.Refs, root, user, email, cookiePath string, env []string) clone.Record {
```
Comment (fejta) [inline prow/clonerefs/run_test.go]: Why can't we change the signature of clone.Run to `refs kube.Refs`?

File: prow/clonerefs/run_test.go [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+	type cloneRec struct {
+		refs        *kube.Refs
+		root        string
+		user, email string
+		cookiePath  string
+		env         []string
+	}
+
+	var recordedClones []cloneRec
```
Comment (BenTheElder) [inline prow/clonerefs/run_test.go]: this has a data race when we write to it from the clone goroutines

**Comment by gyuho** [comment]
@cjwagner Thanks for the fix!

File: prow/clonerefs/run_test.go (hunk 1 above)
Comment (cjwagner) [reply prow/clonerefs/run_test.go]: Done.

File: prow/clonerefs/run_test.go (hunk 2 above)
Comment (cjwagner) [reply prow/clonerefs/run_test.go]: Fixed.

**Comment by BenTheElder** [comment]
Thanks Cole!

**Pull request merged**